internal/ingest/priority_queue.go

```diff
@@ -0,0 +1,95 @@
+package ingest
+
+import "sync"
+
+type Queue struct {
+	lk     sync.Mutex
+	states map[providerID]adInfo
+	order  []providerID
+
+	// used if someone is pulling an empty queue
+	waiters    uint
+	notifyChan chan struct{}
+	doneChan   chan struct{}
+}
+
+func NewPriorityQueue() *Queue {
+	return &Queue{
+		states:     make(map[providerID]adInfo),
+		order:      []providerID{},
```

Comment:
minor: No need to initialize this.  The zero-value of `nil` for this slice will have a `len` of 0 and this will be allocated with the first call to
```go
q.order = append(q.order, p)
```
It works just fine to do: `slice = append(nil, element)`